cpp/src/arrow/dataset/discovery.cc

```diff
@@ -119,24 +114,53 @@ Result<DataSourceDiscoveryPtr> FileSystemDataSourceDiscovery::Make(
               std::move(options));
 }
 
+Result<std::shared_ptr<Schema>> FileSystemDataSourceDiscovery::PartitionSchema() {
+  if (auto partition_scheme = options_.partition_scheme.scheme()) {
+    return partition_scheme->schema();
+  }
+
+  std::vector<util::string_view> paths;
+  for (const auto& stats : forest_.stats()) {
+    if (auto relative =
+            fs::internal::RemoveAncestor(options_.partition_base_dir, stats.path())) {
+      paths.push_back(*relative);
+    }
+  }
+
+  return options_.partition_scheme.discovery()->Inspect(paths);
+}
+
 Result<std::vector<std::shared_ptr<Schema>>>
 FileSystemDataSourceDiscovery::InspectSchemas() {
   std::vector<std::shared_ptr<Schema>> schemas;
 
-  for (const auto& f : files_) {
+  for (const auto& f : forest_.stats()) {
     if (!f.IsFile()) continue;
-    auto source = FileSource(f.path(), fs_.get());
-    ARROW_ASSIGN_OR_RAISE(auto schema, format_->Inspect(source));
+    FileSource src(f.path(), fs_.get());
+    ARROW_ASSIGN_OR_RAISE(auto schema, format_->Inspect(src));
     schemas.push_back(schema);
   }
 
+  ARROW_ASSIGN_OR_RAISE(auto partition_schema, PartitionSchema());
+  schemas.push_back(partition_schema);
+
   return schemas;
 }
 
-Result<DataSourcePtr> FileSystemDataSourceDiscovery::Finish() {
-  ExpressionVector partitions(files_.size(), scalar(true));
+Result<DataSourcePtr> DataSourceDiscovery::Finish() {
+  ARROW_ASSIGN_OR_RAISE(auto schema, Inspect());
+  return Finish(schema);
+}
+
+Result<DataSourcePtr> FileSystemDataSourceDiscovery::Finish(
```

Comment:
The schema handling part of this body should be:

- If a schema is provided, validate it against `Inspect` result.
- If a schema is not provided, use `Inspect` result as the schema.

Think about other DataSource, e.g. OdbcData, how should it validate. The struct options in discovery is class specific, so maybe with functions:

```c++
// Validate infer against provided or use infered schema.
Finish(const std::shared_ptr<Schema>& schema = nullptr);
// Take as-is with no validation, might blow up at runtime.
FinishUnsafe(const std::shared_ptr<Schema>& schema);
```